model/mention_map.go

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"fmt"
+	"net/url"
+)
+
+type UserMentionMap map[string]string
+type ChannelMentionMap map[string]string
+
+const (
+	UserMentionsKey       = "user_mentions"
+	UserMentionsIdsKey    = "user_mentions_ids"
+	ChannelMentionsKey    = "channel_mentions"
+	ChannelMentionsIdsKey = "channel_mentions_ids"
+)
+
+func UserMentionMapFromURLValues(values url.Values) (UserMentionMap, error) {
+	return mentionsFromURLValues(values, UserMentionsKey, UserMentionsIdsKey)
+}
+
+func (m UserMentionMap) ToURLValues() url.Values {
+	return mentionsToURLValues(m, UserMentionsKey, UserMentionsIdsKey)
+}
+func ChannelMentionMapFromURLValues(values url.Values) (ChannelMentionMap, error) {
+	return mentionsFromURLValues(values, ChannelMentionsKey, ChannelMentionsIdsKey)
+}
+
+func (m ChannelMentionMap) ToURLValues() url.Values {
+	return mentionsToURLValues(m, ChannelMentionsKey, ChannelMentionsIdsKey)
+}
+
+func mentionsFromURLValues(values url.Values, mentionKey, idKey string) (map[string]string, error) {
+	mentions, ok := values[mentionKey]
+	if !ok {
+		return nil, fmt.Errorf("%s key not found", mentionKey)
+	}
+
+	ids, ok := values[idKey]
+	if !ok {
+		return nil, fmt.Errorf("%s key not found", idKey)
+	}
+
+	if len(mentions) != len(ids) {
+		return nil, fmt.Errorf("keys %s and %s have different length", mentionKey, idKey)
+	}
+
+	mentionsMap := make(map[string]string)
+	for i, mention := range mentions {
+		mentionsMap[mention] = ids[i]
```

Comment:
As a sanity check, would it make sense to see if `mention` is already in `mentionsMap`, and double check that the `id` hasn't changed?

Reply:
Absolutely! I'll add that case and return an error if it happens.

Reply:
Done in https://github.com/mattermost/mattermost-server/pull/13762/commits/552c67e6d459e98b646fd149c792a2bb0eb7486d.